### PR TITLE
Validate a repair

### PR DESCRIPTION
**Pull request by pudiva**

### CODE REVIEW

File: app/views/repair_requests/new.html.erb [hunk 1]
```diff
@@ -10,58 +10,64 @@
...
+          <% end %>
 
-          <div class="govuk-form-group">
-            <%= form.label :priority, "Task priority", class: 'govuk-label' %>
-            <%= form.collection_select :priority, [['N', '[N] Normal'], ['G', '[G] Great']], :first, :last, {}, class: 'govuk-input' %>
-          </div>
+          <%= form.form_group :priority do %>
+            <%= form.label :priority, "Task priority" %>
+            <%= form.error_messages :priority %>
+            <%= form.collection_select :priority, [['N', '[N] Normal'], ['G', '[G] Great']], :first, :last %>
```
Comment (spannerinworks) [inline app/views/repair_requests/new.html.erb]: where do these vales come from. `[['N', '[N] Normal'], ['G', '[G] Great']]` looks like dummy data

File: app/apis/hackney_api/repairs_client.rb [hunk 2]
```diff
@@ -4,7 +4,14 @@ class RepairsClient
 
     class HackneyApiError < StandardError; end
     class RecordNotFoundError < HackneyApiError; end
-    class ApiError < HackneyApiError; end
+    class ApiError < HackneyApiError;
+      attr_reader :errors
+
+      def initialize msg = nil, errors = nil
```
Comment (spannerinworks) [inline app/apis/hackney_api/repairs_client.rb]: ()

File: app/helpers/govuk_form_helper.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
+module GovukFormHelper
+  class GovukFormBuilder < ActionView::Helpers::FormBuilder
+    def error_messages(key)
+      object.errors[key].map do |msg|
+        @template.content_tag(:span, class: "govuk-error-message") do
+          @template.content_tag(:span, "Error:", class: "govuk-visually-hidden") + @template.sanitize(msg)
```
Comment (spannerinworks) [inline app/helpers/govuk_form_helper.rb:6]: probably should't have nested spans, mozilla says it's not compliant: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/span

File: app/helpers/govuk_form_helper.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    end
+
+    def collection_select(method, collection, value_method, text_method, options = {}, html_options = {})
+      class_and_error_wrap(method, "govuk-select", "govuk-select--error", html_options)
+      super
+    end
+
+    private
+
+    def class_and_error_wrap method, clazz, error_clazz, options
```
Comment (spannerinworks) [inline app/helpers/govuk_form_helper.rb]: ()

File: app/models/hackney/repair_request.rb [hunk 5]
```diff
@@ -63,4 +71,36 @@ def self.attributes_from_api(a)
...
     }
   end
+
+  private
+
+  # TODO: API should give better error descriptions
+  def add_nested_errors
+    errors.each do |key, msg|
+      case key
+      when /^contact\.(.*)$/i
```
Comment (spannerinworks) [inline app/models/hackney/repair_request.rb]: unnecessary `()`

File: app/models/hackney/repair_request.rb [hunk 6]
```diff
@@ -63,4 +71,36 @@ def self.attributes_from_api(a)
...
+        contact.errors.add($1, msg)
+      when /^work_orders\[0\]\.(.*)$/i
+        work_orders[0].errors.add($1, msg)
+      end
+    end
+  end
+
+  # TODO: API should give better error descriptions
+  def self.errors_from_api(e)
```
Comment (spannerinworks) [inline app/models/hackney/repair_request.rb]: better name than `e` pls.

File: app/models/hackney/repair_request.rb [hunk 7]
```diff
@@ -63,4 +71,36 @@ def self.attributes_from_api(a)
...
+    e.map {|x| x["userMessage"] }.group_by do |x|
```
Comment (spannerinworks) [inline app/models/hackney/repair_request.rb]: seems weird that we make strings and then need to parse them with regexes later on

File: app/models/hackney/repair_request.rb [hunk 8]
```diff
@@ -15,6 +15,7 @@ def self.build(attributes)
     new(attributes_from_api(attributes))
   end
 
+  # FIXME: this is view logic and shouldn't be here
```
Comment (spannerinworks) [inline app/models/hackney/repair_request.rb:21]: 👍

**Review by spannerinworks — APPROVED** [review]

File: app/views/repair_requests/new.html.erb (hunk 1 above)
Comment (pudiva) [reply app/views/repair_requests/new.html.erb]: yesss it is. I'm waiting for vlad or bukky to give me a list of priorities.

File: app/helpers/govuk_form_helper.rb (hunk 3 above)
Comment (pudiva) [reply app/helpers/govuk_form_helper.rb:6]: what about this?
https://design-system.service.gov.uk/components/text-input#error-messages

File: app/models/hackney/repair_request.rb (hunk 7 above)
Comment (pudiva) [reply app/models/hackney/repair_request.rb]: indeed this stinks. I just had an idea now that could solve this. Just a sec.

File: app/helpers/govuk_form_helper.rb (hunk 3 above)
Comment (spannerinworks) [reply app/helpers/govuk_form_helper.rb:6]: I trust mozilla more than GDS

Comment (spannerinworks) [reply app/helpers/govuk_form_helper.rb:6]: oh no, it is permitted, never mind: https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content

Comment (pudiva) [reply app/helpers/govuk_form_helper.rb:6]: `<marquee>`

File: app/apis/hackney_api/repairs_client.rb (hunk 2 above)
Comment (pudiva) [reply app/apis/hackney_api/repairs_client.rb]: :heavy_check_mark:

File: app/helpers/govuk_form_helper.rb (hunk 4 above)
Comment (pudiva) [reply app/helpers/govuk_form_helper.rb]: :heavy_check_mark:

File: app/models/hackney/repair_request.rb (hunk 6 above)
Comment (pudiva) [reply app/models/hackney/repair_request.rb]: :heavy_check_mark:

File: app/models/hackney/repair_request.rb (hunk 7 above)
Comment (spannerinworks) [reply app/models/hackney/repair_request.rb]: would it work better as an instance method? Then you could return pairs: `[self.contact, :telephone_number], [self, :description], [work_orders[0], :sor_code]`

File: app/views/repair_requests/new.html.erb (hunk 1 above)
Comment (pudiva) [reply app/views/repair_requests/new.html.erb]: :heavy_check_mark:

**Pull request merged**